app/models/report_template.rb

```diff
@@ -24,6 +24,8 @@ class ReportTemplate < ApplicationRecord
 
   delegate :ability_to_update_for?, to: :session
 
+  before_update :cascade_report_template_type_change
```

Comment:
I think you can use `report_for_changed?` and add this check in this place. For example: 
`before_save :cascade_report_template_type_change, if: :report_for_changed?` (or other trigger)
More info: https://api.rubyonrails.org/classes/ActiveModel/Dirty.html